cloudferrylib/os/identity/keystone.py

```diff
@@ -475,6 +489,35 @@ def _get_user_tenants_roles(self, tenant_list=None, user_list=None):
         if user_list is None:
             user_list = []
         user_tenants_roles = {}
+        if self.config.identity.mode == API_MODE:
+            user_tenants_roles = self.__get_user_tenants_roles_by_api(tenant_list,
+                                                                      user_list)
+        elif self.config.identity.mode == SQL_MODE:
+            user_tenants_roles = self.__get_user_tenants_roles_by_db(tenant_list,
+                                                                     user_list)
+        return user_tenants_roles
+
+    def __get_user_tenants_roles_by_db(self, tenant_list, user_list):
+        user_tenants_roles = {u.name: {t.name: [] for t in tenant_list}
+                              for u in user_list}
+        tenant_ids = {tenant.id: tenant.name for tenant in tenant_list}
+        user_ids = {user.id: user.name for user in user_list}
+        roles = {r.id: r for r in self.get_roles_list()}
+        res = self.mysql_connector.execute(
+            "SELECT * FROM user_project_metadata")
```

Comment:
Can this request possibly have incorrect data? For example, will it include deleted roles, or deleted tenants, something we may fail on?